Adding bitbucket cloud support

gcase555 opened the pull request:
# What does this change
This is most of the changes for adding bitbucket cloud support. We have implemented the interface and have the ability to automate creating PRs, updating PRs, Closing PRs, Fetching Repos/PRs and Forking Repos.

# What issue does it fix
Closes #99


# Notes for the reviewer
We are still adding support for handling multiple workspaces and updating the README with a few notes for bitbucket cloud support. My goal here is to get the PR submitted for early feedback and we will update this PR once we add handling multiple workspaces.

gcase555 commented on examples/go/upgrade-go-version.sh:
```diff
@@ -2,5 +2,5 @@
 # Title: Upgrade Go version in go modules
 
-go mod edit -go 1.18
+go mod edit -go 1.19
```
We changed this for testing, we'll switch it back

gcase555 commented on internal/scm/bitbucketcloud/custom_structs.go:
```diff
@@ -0,0 +1,109 @@
+	stateDeclined = "DECLINED"
+)
+
+type bitbucketPullRequests struct {
```
need to fix the formatting here

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+		Owner:    bbc.workspaces[0],
+		RepoSlug: splitRepoFullName[1],
+	}
+	defaultReviewers, _ := bbc.bbClient.Repositories.Repository.ListDefaultReviewers(repoOptions)
```
err check?

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+		newPR.Reviewers = append(newPR.Reviewers, reviewer.Uuid)
+	}
+
+	prOptions := &bitbucket.PullRequestsOptions{
```
Can pull request happen from a fork in bbc? In that case I don't think the owner here is correct?

gcase555 replied:
We need to add a parameter to allow users to specify a bitbucket "project" to fork into before we can specify the owner fully here. 

We were hoping to do that as a separate PR and document that as a known limitation but let us know if that is a blocker we should resolve it before we get this change merged.

lindell replied:
I'm fine with merging something that does not have all functionality of all other systems 🙂 But please list these, as I will otherwise make comments on things that will not be relevant to you

gcase555 replied:
I made a note in the README template about bitbucket cloud support and the current limitations that should help clarify things.

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+		project:    splitRepoFullName[0],
+		repoName:   splitRepoFullName[1],
+		branchName: newPR.Head,
+		prProject:  splitRepoFullName[0],
```
It seems that `project`/`repoName` and `prProject`/`prRepoName` are used the same here. I guess it comes from other implementations and it might be good to clarify what they are for there.

The first pair are a reference to the project where the PR is place (where is will merge), while the second one is from which repository the change is currently made. If `fork` is not defined, these will be the same, since the changes will happen the same project. But they could be different, and in that case I think this will not be right.

lindell reviewed (commented):
Thanks for the PR 😄 

I started looking at it, but it will take some time since I don't have a lot of time currently. But I'm including my first comments. Feel free to take a look at them right away if you want, or wait until all comments are in if you prefer that 🙂 

PS. In which detail do you want the reviews right now? I tried to avoid code style, and similar types of comments because of the current PR state you mentioned. But let me know if you prefer those straight away as well.

gcase555 commented on cmd/platform.go:
```diff
@@ -283,6 +286,35 @@ func createGiteaClient(flag *flag.FlagSet, verifyFlags bool) (multigitter.Versio
+		return nil, errors.New("no workspace, user or repository set")
+	}
+
+	if username == "" {
```
We initially tried this but the problem is most OAuth functionality is gated behind bitbucket premium, app passwords are what is most accessible for bitbucket users and are going to support it first before we add OAuth support since many bitbucket users are not premium.

The problem is the amount of permissions multi-gitter needs, in order to access multiple projects or workspaces you need a higher permission token and repository tokens(that are free) are far too limited to handle updating multiple repositories in a project or workspace.

gcase555 replied:
App passwords are what is recommended over OAuth tokens for non-premium users, they have some similar features including the ability to limit access to what an app password can do and the ability to revoke and re-generate easily.

We would like to add OAuth support as an additional feature in a separate PR to keep the scope of this change small as well.

lindell replied:
SGTM 👍

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+func (bbc *BitbucketCloud) CreatePullRequest(ctx context.Context, repo scm.Repository, prRepo scm.Repository, newPR scm.NewPullRequest) (scm.PullRequest, error) {
+
+	splitRepoFullName := strings.Split(prRepo.FullName(), "/")
```
Could you calrify what you mean by `type asserting` here?

lindell replied:
`prRepo.(repository)` is type asserting to the repository struct. Since it's guaranteed to be of that type, we can do it. The important part is to make it clearer

`splitRepoFullName[1]` is hard to understand without first parsing the surrounding code.
`bbcPrRepo.name` is self explanatory 😄

gcase555 replied:
Type assertions came in handy and we were mostly able to use them.

gcase555 commented:
> PS. In which detail do you want the reviews right now?

The feedback you have given so far has been good, we were looking to answer things that stood out and could potentially be problems or explain why we had to go with app passwords over OAuth access tokens for this first pull-request for accessibility reasons.

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+func (bbc *BitbucketCloud) GetPullRequests(ctx context.Context, branchName string) ([]scm.PullRequest, error) {
+	var responsePRs []scm.PullRequest
+	for _, repoName := range bbc.repositories {
+		prs, _ := bbc.bbClient.Repositories.PullRequests.Gets(&bitbucket.PullRequestsOptions{Owner: bbc.workspaces[0], RepoSlug: repoName})
+		prBytes, _ := json.Marshal(prs)
+		bbPullRequests := &bitbucketPullRequests{}
+		err := json.Unmarshal(prBytes, bbPullRequests)
+		if err != nil {
```
I find it strange that the bbc package does return `any` (`interface{}`) and not concrete types. That should be one of the biggest benefits of using a pre defined package instead of rolling your own. But let us work under this constraint. 

Since we have to do the marshal->unmarshal in multiple places, there would be benefit in creating a helper function for this

```go
func decode(raw any, structured any) error {
	prBytes, err := json.Marshal(prs)
    if err != nil { // I guess this should never happen, since reference loops etc should not exist, but does not hurt to check it
        return err 
    }
    return json.Unmarshal(prBytes, structured)
}
```

An alternative is to use https://github.com/mitchellh/mapstructure , but I'm fine if you want to do go the json approach since that is where it is unmarshaled from to begin with.

gcase555 replied:
We only had to do the marshal/unmarshal a few times so I decided not to add the decode function, I don't think its excessively repetitive to keep it as is but let me know if you think its blocking

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+func (bbc *BitbucketCloud) GetPullRequests(ctx context.Context, branchName string) ([]scm.PullRequest, error) {
+	var responsePRs []scm.PullRequest
+	for _, repoName := range bbc.repositories {
+		prs, _ := bbc.bbClient.Repositories.PullRequests.Gets(&bitbucket.PullRequestsOptions{Owner: bbc.workspaces[0], RepoSlug: repoName})
```
errcheck

lindell commented on internal/scm/bitbucketcloud/custom_structs.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+// repository contains information about a bitbucket repository
+type repository struct {
```
Adding workspace name is probably a good idea for later, even if multi workspace support is currently not implemented.

gcase555 replied:
Could this be handled later? I didn't want to refactor too much more and were still only supporting single workspaces(as documented in the README)

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+	return err
+}
+
+func (bbc *BitbucketCloud) GetRepositories(ctx context.Context) ([]scm.Repository, error) {
```
This seems to only include repository name filtering (`--repo`), not for example per project which I think is by far the most popular one (organizations for GitHub).

And for specific repo names, it seems that all repositories are fetched. If possible, I think getting those repositories by name should be prefered.

gcase555 replied:
I don't think we had time to implement this change, is it ok if we keep this as is with repository filtering? Trying to avoid heavier refactoring of this PR as it has gotten large and would prefer to do it as a follow up

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+func (bbc *BitbucketCloud) ForkRepository(ctx context.Context, repo scm.Repository, newOwner string) (scm.Repository, error) {
+	splitRepoFullName := strings.Split(repo.FullName(), "/")
```
Same comment as before.

lindell commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+	_ = json.Unmarshal(linkBytes, rLinks)
+
+	if bbc.sshAuth {
+		cloneURL = findLinkType(rLinks.Clone, cloneSSHType)
+		if cloneURL == "" {
+			return nil, errors.Errorf("unable to find clone url for repository %s using clone type %s", repo.Name, cloneSSHType)
+		}
```
Move the error to inside the `findLinkType` function, and let it return an error

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+	}, nil
+}
+
+func (bbc *BitbucketCloud) GetPullRequests(ctx context.Context, branchName string) ([]scm.PullRequest, error) {
```
Good catch, I added the branch name

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+		return nil, err
+	}
+
+	return &pullRequest{
```
Good catch, we added ID and guiURL

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+	return scm.PullRequestStatusSuccess, nil
+}
+
+func (bbc *BitbucketCloud) GetOpenPullRequest(ctx context.Context, repo scm.Repository, branchName string) (scm.PullRequest, error) {
```
I think thats ok for us if its not used by multi-gitter, its just how we created the method

gcase555 replied:
Wait I am slightly confused, this _is_ part of the version controller interface https://github.com/lindell/multi-gitter/blob/master/internal/multigitter/run.go#L363

gcase555 replied:
To be clear we just implemented the interface, this is part of it

gcase555 commented on docs/README.template.md:
```diff
@@ -132,3 +132,27 @@ All configuration in multi-gitter can be done through command line flags, config
 {{end}}{{end}}
 
 Do you have a nice script that might be useful to others? Please create a PR that adds it to the [examples folder](/examples).
+
```
@lindell we added a section here in the README to document how to use bitbucket cloud and the limitations as well as links to official documentation.

gcase555 replied:
done, comments added about beta support and documented any issues/missing features

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+		branchName: newPR.Head,
+		prProject:  splitRepoFullName[0],
+		prRepoName: splitRepoFullName[1],
+		status:     scm.PullRequestStatusSuccess,
```
`bbc.pullRequestStatus()` required a parameter to be passed in that I didn't have available in this function so we went with `scm.PullRequestStatusSuccess` to keep it simple

gcase555 replied:
done

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,324 @@
+
+func (bbc *BitbucketCloud) MergePullRequest(ctx context.Context, pr scm.PullRequest) error {
+	bbcPR := pr.(pullRequest)
+	repoSlug := strings.Split(bbcPR.guiURL, "/")
```
Its not the same as `repoSlug[4]`, I used a more clearly named variable instead but still parsed it out of the guiUrl. If its ok I would prefer to not refactor further as this PR is getting large and instead use the variable and parsed slug name using the slice index.

gcase555 commented:
@lindell We made as many of the changes as we could with the time we had allotted to work on this and did most of the cleanup and refactoring requested. Please let us know if there are any issues that are blocking and _need_ to be addressed before merging or if there is something we can resolve in a follow up pull request, we are hoping to get this merged as this PR has started getting large and has been outside of master branch for awhile.

I think this implementation covers enough to close out issue #99 and can start being used by end users. We could create a separate issue to handle updates to include project settings and addressing other smaller feedback tasks you feel are non-blocking.

We have successfully started using it to update multiple services across large and small bitbucket cloud workspaces and included documentation on how to get it working and the current limitations.

lindell reviewed (commented):
As per the pervious comment, I did a pass and tried to be non-nitty.

Just to clarify, you are employees of Atlassian? It does change the calculation a bit, since you would have greater investment in the project after the merge. 

I think the best approach right with your wish to have i merged in the current state is:
1) Go through the last comments + Fix the linting errors (feel free to skip the G601 ones, since it will be fixed when we update to Go 1.22)
2) Mark the project as Beta in the Readme and in the description of the `--platform` flag/option.

waded commented:
I tested this, and was successful in creating some PRs across multiple repositories. Love it! 

Two areas for improvement:

The error when `--org` isn't set, `panic: runtime error: index out of range [0] with length 0`, isn't intuitive. If it's required it'd be nice if the error explained this. IDK if this is an issue w/ all platforms, or is specific to bitbucket_cloud, but it was the first issue I ran into.

We have a LOT of repositories. Although I provided fewer than 10 to `--repo` I did receive a `could not fetch repositories: 429 Too Many Requests` after a few dry runs. Is there perhaps something non-optimal about the number of calls its making to Bitbucket Cloud?

gcase555 commented on internal/scm/bitbucketcloud/bitbucket_cloud.go:
```diff
@@ -0,0 +1,373 @@
+
+func (bbc *BitbucketCloud) MergePullRequest(ctx context.Context, pr scm.PullRequest) error {
+	bbcPR := pr.(pullRequest)
+	repoSlug := strings.Split(bbcPR.guiURL, "/")[4]
```
I mentioned it in a resolved comment but no, when I tried to use `bbcPR.repoName` it did not work and this slug is different. I used a clear variable name to clear up confusion

gcase555 replied:
The previous 4 parts are the complete URL but for the repo slug we only care about the repo name which we are extracting from the URL in this case

gcase555 commented:
> As per the pervious comment, I did a pass and tried to be non-nitty.
> 
> Just to clarify, you are employees of Atlassian? It does change the calculation a bit, since you would have greater investment in the project after the merge.

Correct, we are developers from Atlassian and we use golang heavily on some platform tools we released here https://github.com/asecurityteam/ we still plan on submitting some smaller follow up PRs but we are trying to get this merged with the working functionality in place since this PR is getting to large and been open for longer then we prefer.

> I think the best approach right with your wish to have i merged in the current state is:
> 
>     1. Go through the last comments + Fix the linting errors (feel free to skip the G601 ones, since it will be fixed when we update to Go 1.22)
> 
>     2. Mark the project as Beta in the Readme and in the description of the `--platform` flag/option.

I addressed all the nits, linter fixes and added a note on Beta support in the README in our fork here https://github.com/asecurityteam/multi-gitter/pull/7 I plan on requesting a re-reviewing once I get that merged.

gcase555 commented:
> I tested this, and was successful in creating some PRs across multiple repositories. Love it!

I am really glad to hear that!

> The error when --org isn't set, panic: runtime error: index out of range [0] with length 0, isn't intuitive. If it's required it'd be nice if the error explained this. I'm new to multi-gitter, and so unsure if this is an issue w/ all platforms, or is specific to bitbucket_cloud.

Noted, this is something we will probably improve in a small follow up PR

> Two areas for improvement:
> We have 1000s of repositories in Bitbucket Cloud. Although I provided the names of fewer than 10 to `--repo` I did receive a `could not fetch repositories: 429 Too Many Requests` after a few dry runs (which each took far longer than I expected.) After letting things cool down, got past the 429, but still it was pretty slow. The 10 repos in question are very small. Is there perhaps something non-optimal about the number of calls it makes to Bitbucket Cloud?

We did make a note in the README that workspaces with lots of repositories will perform slower, it will still work but there is usually a delay when initially doing look ups for repos and PRs, you should still see commands execute within 2 minutes. We noted this as part of the Beta support that it does work but the performance will likely get better once we add more options for project support. With smaller workspaces, it runs very fast so this is something specific to large workspaces.

gcase555 commented:
Alright ready for a re-review, we addressed all linting issues besides the ones you mentioned we could ignore, did the requested cleanup of TODOs, added the function for extracting the repo slug, made notes on beta support and updated the docs to use details to make them more minimal.

gcase555 commented:
@lindell is it possible we could get help merging this?

lindell commented:
I'm about to merge this PR. But I can push the go mod changes necessary (fixes for merge conflicts).

It seems that "Maintainers are allowed to edit this pull request." is enabled, but I get 
````
ERROR: Permission to asecurityteam/multi-gitter.git denied to lindell.
````

When I try to push the changes.

kpocius commented:
> I'm about to merge this PR. But I can push the go mod changes necessary (fixes for merge conflicts).
> 
> It seems that "Maintainers are allowed to edit this pull request." is enabled, but I get
> 
> ```
> ERROR: Permission to asecurityteam/multi-gitter.git denied to lindell.
> ```
> 
> When I try to push the changes.

This is a known limitation of GitHub. That setting doesn't work if the repository belongs to an organization.

https://github.com/orgs/community/discussions/5634

gcase555 commented:
Ill address the merge conflicts now

gcase555 commented:
@lindell all merge conflicts have been addressed

lindell commented:
Thanks for the contribution. I am looking forward to the fixes coming soon 🙂
